### PR TITLE
MGMT-4762 Support HEAD for ISO download

### PR DESCRIPTION
**Pull request by mmartinv**

### CODE REVIEW

File: swagger.yaml [hunk 1]
```diff
@@ -392,6 +392,53 @@ paths:
           schema:
             $ref: '#/definitions/error'
 
+    head:
+      tags:
+        - installer
+      description: Downloads the OpenShift per-cluster Discovery ISO Headers only.
+      security:
+        - userAuth: [admin, read-only-admin, user]
```
Comment (avishayt) [inline swagger.yaml:400]: Under this line add also:
- urlAuth: []

File: swagger.yaml [hunk 2]
```diff
@@ -392,6 +392,53 @@ paths:
...
+      operationId: DownloadClusterISOHeaders
+      produces:
+        # application/vnd.efi.iso is not supported
+        - application/octet-stream
```
Comment (avishayt) [inline swagger.yaml:404]:
```suggestion
```

File: swagger.yaml [hunk 3]
```diff
@@ -392,6 +392,53 @@ paths:
...
+      parameters:
+        - in: path
+          name: cluster_id
+          description: The cluster whose ISO headers should be downloaded.
+          type: string
+          format: uuid
+          required: true
+      responses:
+        "200":
+          description: Success.
```
Comment (avishayt) [inline swagger.yaml:414]:
```suggestion
          description: Success.
          headers:
            Content-Length:
              type: integer
              description: Size of the ISO in bytes
```

File: internal/bminventory/inventory.go [hunk 4]
```diff
@@ -711,6 +711,47 @@ func (b *bareMetalInventory) DownloadClusterISO(ctx context.Context, params inst
...
+	if err != nil {
+		log.WithError(err).Errorf("Failed to get ISO for cluster %s", cluster.ID.String())
+		b.eventsHandler.AddEvent(ctx, params.ClusterID, nil, models.EventSeverityError,
+			"Failed to download image: error fetching from storage backend", time.Now())
+		return installer.NewDownloadClusterISOHeadersInternalServerError().
+			WithPayload(common.GenerateError(http.StatusInternalServerError, err))
+	}
+	if !exists {
+		b.eventsHandler.AddEvent(ctx, params.ClusterID, nil, models.EventSeverityError,
+			"Failed to download image: the image was not found (perhaps it expired) - please generate the image and try again", time.Now())
```
Comment (avishayt) [inline internal/bminventory/inventory.go:734]: No need for this event

File: internal/bminventory/inventory.go [hunk 5]
```diff
@@ -711,6 +711,47 @@ func (b *bareMetalInventory) DownloadClusterISO(ctx context.Context, params inst
...
+		return installer.NewDownloadClusterISOHeadersNotFound().
+			WithPayload(common.GenerateError(http.StatusNotFound, errors.New("The image was not found "+
+				"(perhaps it expired) - please generate the image and try again")))
```
Comment (avishayt) [inline internal/bminventory/inventory.go:737]:
```suggestion
			WithPayload(common.GenerateError(http.StatusNotFound, errors.New("The image was not found")))
```

File: internal/bminventory/inventory.go [hunk 6]
```diff
@@ -711,6 +711,47 @@ func (b *bareMetalInventory) DownloadClusterISO(ctx context.Context, params inst
...
+	_, contentLength, err := b.objectHandler.Download(ctx, imgName)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to get ISO for cluster %s", cluster.ID.String())
+		b.eventsHandler.AddEvent(ctx, params.ClusterID, nil, models.EventSeverityError,
+			"Failed to download image: error fetching from storage backend", time.Now())
+		return installer.NewDownloadClusterISOHeadersInternalServerError().
+			WithPayload(common.GenerateError(http.StatusInternalServerError, err))
+	}
+	b.eventsHandler.AddEvent(ctx, params.ClusterID, nil, models.EventSeverityInfo,
+		fmt.Sprintf(`Started image download (image type is "%s")`, cluster.ImageInfo.Type), time.Now())
```
Comment (avishayt) [inline internal/bminventory/inventory.go:748]:
```suggestion
	imgSize, err := b.objectHandler.GetObjectSizeBytes(ctx, imgName)
	if err != nil {
		log.WithError(err).Errorf("Failed to get ISO size for cluster %s", cluster.ID.String())
		return common.NewApiError(http.StatusBadRequest, err)
	}
```

File: internal/bminventory/inventory.go [hunk 7]
```diff
@@ -711,6 +711,47 @@ func (b *bareMetalInventory) DownloadClusterISO(ctx context.Context, params inst
...
+
+	return filemiddleware.NewResponder(installer.NewDownloadClusterISOHeadersOK(),
+		fmt.Sprintf("cluster-%s-discovery.iso", params.ClusterID.String()),
+		contentLength)
```
Comment (avishayt) [inline internal/bminventory/inventory.go:752]: Fix this to return 200 and Content-Length (use the method WithContentLength)

File: internal/bminventory/inventory.go [hunk 8]
```diff
@@ -711,6 +711,47 @@ func (b *bareMetalInventory) DownloadClusterISO(ctx context.Context, params inst
 		contentLength)
 }
 
+func (b *bareMetalInventory) DownloadClusterISOHeaders(ctx context.Context, params installer.DownloadClusterISOHeadersParams) middleware.Responder {
```
Comment (avishayt) [inline internal/bminventory/inventory.go:714]: Unit tests?

**Review by avishayt — CHANGES_REQUESTED** [review]

**Comment by carbonin** [comment]
/approve

Looks good. I think you have a linting error though.
Make sure you're generating the files using the generator make targets. I think you'll want `skipper make generate-from-swagger` or maybe just go with `skipper make generate-all`

**Comment by avishayt** [comment]
Can you please show the output of the API (specifically the content length) when using a full ISO and a minimal ISO?  Let's make sure it's as expected.  Thanks!

**Review by rwsu — COMMENTED** [review]
I would suggest adding to the commit message with something like: "Adds content-length header for ISO downloads. Content-length header and status 200 are required by Ironic."

**Comment by flaper87** [comment]
@mmartinv hope you don't mind but I took over this PR and pushed the fixes for CI here https://github.com/openshift/assisted-service/pull/1430

I also addressed the comments from @avishayt and @rwsu 

I think we can close this one :)

**Pull request closed without merging**